AdaGrad Update Policy

abhinavmoudgil95 opened the pull request:
__Implementation of AdaGrad optimizer__ 
* Added AdaGrad update policy class to SGD. 
* All tests passed. 
* Detailed discussion for reducing code duplication in #897. 
* Reference: [Adaptive Subgradient Methods for Online Learning and Stochastic Optimization](http://jmlr.org/papers/v12/duchi11a.html).

abhinavmoudgil95 commented:
@rcurtin @zoq PR is ready to be merged. Please suggest edits, if any.

abhinavmoudgil95 commented:
@rcurtin @zoq PR is ready to be merged. Please review and suggest edits, if any.

zoq commented on src/mlpack/core/optimizers/ada_grad/ada_grad.hpp:
```diff
@@ -0,0 +1,172 @@
+
+  //! Stochastic Gradient Descent object with AdaGrad policy.
+  SGD<DecomposableFunctionType, AdaGradUpdate>
+  optimizer(DecomposableFunctionType&,
```
I'm wondering if this works for you? Shouldn't that be ``SGD<DecomposableFunctionType, AdaGradUpdate> optimizer;``?

abhinavmoudgil95 replied:
Yes, I have made the change. I was just testing a few things to resolve build errors. 

Actually, I am not able to build locally as CMake gets stuck after executing few steps when I run command `cmake ..` (details [here](https://paste.debian.net/hidden/e457f447/)). Could you give an idea why it might be happening?

abhinavmoudgil95 replied:
Yes, I have made the change. I was just testing a few things to resolve build errors. 

Actually, I am not able to build locally now as CMake gets stuck after executing few steps when I run command `cmake ..` (details [here](https://paste.debian.net/hidden/e457f447/)). Earlier when I opened up AdaGrad PR, I was able to build and test. Could you give an idea why it might be happening?

rcurtin replied:
The link doesn't work, so I can't provide any help.  Are you still having the issue?

abhinavmoudgil95 replied:
Yes, issue is still there. Sorry, the link got expired. This is the updated [link](http://pastebin.com/tGFjGdmk). I work on a GPU server where I don't have root access. Hence, I updated the `cmake` command accordingly. It remains stuck after the last line on console:
` Found Doxygen: /home/abhinav.moudgil/doxygen_installed/bin/doxygen (found version "1.8.14") `.

zoq commented on src/mlpack/tests/ada_grad_test.cpp:
```diff
@@ -0,0 +1,107 @@
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#include <mlpack/core.hpp>
+#include <mlpack/core/optimizers/sgd/sgd.hpp>
```
I think we can remove the header file here. We include ``SGDTestFunction`` with ``#include <mlpack/core/optimizers/sgd/test_function.hpp>``.

abhinavmoudgil95 replied:
Done.

zoq commented on src/mlpack/core/optimizers/sgd/update_policies/CMakeLists.txt:
```diff
@@ -1,6 +1,7 @@
 set(SOURCES
   vanilla_update.hpp
   momentum_update.hpp
+  ada_grad_update.hpp
```
I looks like we moved the file from ``/core/optimizers/sgd/update_policies/``to the adagrad folder, so we should remove the entry from the sgd policies CMakeLists.txt file.

abhinavmoudgil95 replied:
Fixed.

abhinavmoudgil95 replied:
Done.

zoq commented on src/mlpack/core/optimizers/ada_grad/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+set(SOURCES
+  ada_grad.hpp
```
We should include the update policy here.

abhinavmoudgil95 replied:
Fixed.

abhinavmoudgil95 commented:
@zoq Thanks a lot for the review. Sorry, I got busy in my mid term examinations, so couldn't update the PR. I am making changes now.

rcurtin commented on src/mlpack/core/optimizers/ada_grad/ada_grad.hpp:
```diff
@@ -0,0 +1,164 @@
+
+  //! Controls whether or not the individual functions are shuffled when
+  //! iterating.
+  bool shuffle;
```
Why do all these parameters need to be held in the class if they are not used and are also held in `optimizer`?  I think we could remove them.

abhinavmoudgil95 replied:
Yes, I think we can remove all the parameters as they are held by optimizer. Just not sure about `adagradUpdate`.

rcurtin commented on src/mlpack/core/optimizers/ada_grad/ada_grad.hpp:
```diff
@@ -0,0 +1,164 @@
+  bool shuffle;
+
+  //! The AdaGradUpdate policy object.
+  AdaGradUpdate adagradUpdate;
```
The update policy is also held in `optimizer`, so I think we could remove this member too?

zoq replied:
I think it's need for ``double Epsilon() const { return adagradUpdate.Epsilon(); }`` but that shouldn't have an effect on the optimizer class since it uses: 

```
const UpdatePolicy updatePolicy = UpdatePolicy());
```

abhinavmoudgil95 replied:
Can we change the `Epsilon()` function to something like below? I am not sure about this but I feel that `UpdatePolicy` is also part of SGD, so it should work. 
```c++
 double Epsilon() const { return optimizer.AdaGradUpdate.Epsilon(); }
```

Accordingly, I would change the constructor: 

```c++
AdaGrad<DecomposableFunctionType>::AdaGrad(DecomposableFunctionType& function,
                                           const double stepSize,
                                           const double epsilon,
                                           const size_t maxIterations,
                                           const double tolerance,
                                           const bool shuffle)
{
  AdaGradUpdate adagradUpdate(epsilon);
  optimizer = SGD<DecomposableFunctionType, AdaGradUpdate>(function, stepSize, 
         maxIterations, tolerance, shuffle, adagradUpdate);
}
```

rcurtin replied:
Yes, I think it would be `optimizer.UpdatePolicy().Epsilon()` but that should work and would be a good way to do it without duplicating any options anywhere.

abhinavmoudgil95 replied:
Ah thanks, I will make the changes asap.

zoq replied:
``optimizer.UpdatePolicy().Epsilon()`` That only works if SGD would implement a function that returns the UpdatePolicy. So I guess we either have to implement that function or we could change the constructor to use a reference for the policy.

abhinavmoudgil95 replied:
Keeping consistency in mind, I think implementing a function `UpdatePolicy()` just like other variables would be better?

rcurtin replied:
I agree, I am sure that `UpdatePolicy()` will come in useful in other situations too.

abhinavmoudgil95 commented:
@rcurtin @zoq I am facing some issues in removing `adagradUpdate` parameter from AdaGrad class. 

We are declaring the SGD optimizer in AdaGrad class without any arguments like `SGD<DecomposableFunctionType, AdaGradUpdate> optimizer;` whereas the default constructor of SGD has 6 arguments. To counter this, we must initialize the `optimizer` in AdaGrad constructor with initializer list like below otherwise it doesn't work as default constructor expects 6 arguments and we are passing 0 in the declaration.
```c++
AdaGrad<DecomposableFunctionType>::AdaGrad(DecomposableFunctionType& function,
                                           const double stepSize,
                                           const double epsilon,
                                           const size_t maxIterations,
                                           const double tolerance,
                                           const bool shuffle):
adagradUpdate(epsilon),
optimizer(function, stepSize ...) // requires AdaGradUpdate object
{/* Nothing to do */}
```
But initializing `optimizer` in intializer list like above requires the `AdaGradUpdate` object which we don't have.

We can resolve this by adding a constructor without any arguments `SGD();` to SGD class. Then I can reinitialize my `optimizer` object in AdaGrad constructor as I suggested before:
```c++
AdaGrad<DecomposableFunctionType>::AdaGrad(DecomposableFunctionType& function,
                                           const double stepSize,
                                           const double epsilon,
                                           const size_t maxIterations,
                                           const double tolerance,
                                           const bool shuffle)
{
  AdaGradUpdate adagradUpdate(epsilon);
  optimizer = SGD<DecomposableFunctionType, AdaGradUpdate>(function, stepSize, 
         maxIterations, tolerance, shuffle, adagradUpdate);
}
``` 
Please suggest if there is any other better way to solve this problem.

mlpack-jenkins commented:
Can one of the admins verify this patch?

abhinavmoudgil95 commented:
* Removed all the parameters from AdaGrad class except `SGD` and `AdaGradUpdate` object. 
* Added `GetUpdatePolicy()` method to SGD to access `UpdatePolicy` object. Then we can access `epsilon` directly from optimizer rather than accessing it from `AdaGradUpdate` object. 
* Had to name `GetUpdatePolicy()` rather than `UpdatePolicy()` otherwise template parameter gets shadowed and it doesn't work. Also, it can't be overloaded `UpdatePolicy& GetUpdatePolicy()` like other parameter methods.

zoq commented on src/mlpack/core/optimizers/sgd/sgd.hpp:
```diff
@@ -135,6 +136,9 @@ class SGD
   bool& Shuffle() { return shuffle; }
 
+  //! Get the update policy. 
+  UpdatePolicy GetUpdatePolicy() const { return updatePolicy; }
```
Instead of using ``GetUpdatePolicy`` because of the template parameter shadowed issue, you can just rename the template parameter to ``updatePolicyType`` that way we are still consistent with the rest of the codebase.

zoq replied:
Also please add ``UpdatePolicy& UpdatePolicy() { return updatePolicy; }``, so that we can modify the policy.

abhinavmoudgil95 replied:
That's a good idea. So, I should make the following updates in SGD class, right? 
* `updatePolicy` to `updatePolicyType`
* `typename UpdatePolicy = VanillaUpdate` to `typename UpdatePolicyType = VanillaUpdate`
* `GetUpdatePolicy()` to `UpdatePolicy()`

zoq replied:
yes, but please only rename the type not the parameter so:

```
const UpdatePolicy updatePolicy = UpdatePolicy());
```

would become 

```
const UpdatePolicyType updatePolicy = UpdatePolicyType());
```

and not

```
const UpdatePolicyType updatePolicyType = UpdatePolicyType());
```

and also don't forget to add:

```
UpdatePolicyType& UpdatePolicy() { return updatePolicy; }
```

to the sgd class.

abhinavmoudgil95 replied:
Sure, thanks a lot.

rcurtin commented on src/mlpack/core/optimizers/ada_grad/ada_grad.hpp:
```diff
@@ -0,0 +1,145 @@
+
+ private:
+  //! The AdaGradUpdate policy object.
+  AdaGradUpdate adagradUpdate;
```
This member isn't used and I think we should remove it---there is already one of these held inside `optimizer`.

rcurtin commented on src/mlpack/core/optimizers/ada_grad/ada_grad_impl.hpp:
```diff
@@ -0,0 +1,39 @@
+              maxIterations,
+              tolerance,
+              shuffle,
+              adagradUpdate)
```
You can replace `adagradUpdate` with `AdagradUpdate(epsilon)`.

rcurtin commented:
Looks good to me, although since Travis and AppVeyor refuse to build since there are merge conflicts, it should be tested during the merge process.  @zoq: I can do that if you want and merge this in, just let me know.

rcurtin commented:
Also, Abhinav, thanks for the hard work and patience with this. :)  I know there were lots of changes and it ended up looking way different than the original version.

zoq commented:
Let's check the code one more time :)

abhinavmoudgil95 commented:
Thanks a lot, Ryan and Marcus for guiding me so well. Honestly, it has been a great learning experience for me. :)

zoq commented:
Thanks again for the contribution.

abhinavmoudgil95 commented:
Glad to see the PR merged. Really excited to contribute more!
